File: clang/dynamfib.c
/* File: dynamfib.c
 * Name: D.Saravanan
 * Date: 22/11/2022
 * Program to compute nth fibonacci using dynamic programming
*/

#include <stdio.h>
#include <gmp.h>


mpz_t *fibonacci(mpz_t * result, int n) {
    /* compute nth fibonacci */

    mpz_t fib[2];

    mpz_init_set_ui(fib[0], 0);
    mpz_init_set_ui(fib[1], 1);

    for (size_t i = 0; i < n; i++) {
	mpz_add(fib[0], fib[0], fib[1]);
	mpz_swap(fib[0], fib[1]);
    }

    mpz_set(*result, fib[0]);
    return result;
}


int main() {

    int N = 10000;

    mpz_t result;
    mpz_init(result);

    gmp_printf("%Zd\n", fibonacci(&result, N));

    return 0;
}

File: golang/dynamfib.go
// File: dynamfib.go
// Name: D.Saravanan
// Date: 29/11/2022
// Program to compute nth fibonacci using dynamic programming

package main

import "fmt"
import "math/big"

func fibonacci(n int) *big.Int {
	/* compute nth fibonacci */

	fib := [2]*big.Int{big.NewInt(0), big.NewInt(1)}

	for i := 0; i < n; i++ {
		fib[0].Add(fib[0], fib[1])
		fib[0], fib[1] = fib[1], fib[0]
	}

	return fib[0]
}

func main() {
	N := 10000
	fmt.Println(fibonacci(N))
}
